Make web process send ShareableBitmap instead of SharedBuffer for icon data

szewai opened the pull request:
#### db75e4455949f92e7e15b3ca1514466d66811788
<pre>
Make web process send ShareableBitmap instead of SharedBuffer for icon data
<a href="https://bugs.webkit.org/show_bug.cgi?id=290873">https://bugs.webkit.org/show_bug.cgi?id=290873</a>
<a href="https://rdar.apple.com/143579665">rdar://143579665</a>

Reviewed by Said Abou-Hallawa and Anne van Kesteren.

In existing implementation of icon data generation, UI process sends image data to web process, and web process decodes
image from data, generates images with different sizes, combines them into one ico image and sends back image data to UI
process. To make the process more safe, this patch makes web process send images (bitmaps) with different sizes to UI
process, and UI process is responsible for combining them into an ico image.

* Source/WebCore/platform/graphics/ImageAdapter.h:
(WebCore::ImageAdapter::image const):
* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::createFromImageDraw):
* Source/WebCore/platform/graphics/ShareableBitmap.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createBitmapsFromNativeImage):
(WebCore::createNativeImageFromSVGImage):
(WebCore::createBitmapsFromSVGImage):
(WebCore::createBitmapsFromImageData):
(WebCore::createIconDataFromBitmaps):
(WebCore::expandNativeImageToData): Deleted.
(WebCore::expandSVGImageToData): Deleted.
(WebCore::createIconDataFromImageData): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createIconDataFromImageData):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createBitmapsFromImageData):
(WebKit::WebPage::createIconDataFromImageData): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/icon-svg-16.tiff: Added.
* Tools/TestWebKitAPI/Tests/WebKit/icon-svg-256.tiff: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::tiffRepresentation):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVG)):

Canonical link: <a href="https://commits.webkit.org/293430@main">https://commits.webkit.org/293430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b248879bd902c642dd58ac2f964a96bfe63a4968

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98889 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104015 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49478 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100934 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18821 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26976 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75282 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32413 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101893 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89306 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55643 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/14090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7278 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48858 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84027 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7351 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106384 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25986 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18942 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84246 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85504 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83744 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21226 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28399 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6071 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19708 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25939 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31125 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25759 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29079 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27333 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,87 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-        CGImageDestinationAddImage(destination.get(), newImage->platformImage().get(), nullptr);
+        IntSize size { (int)length, (int)length };
+        if (RefPtr nativeImage = image.nativeImage(image.size())) {
+            if (RefPtr bitmap = ShareableBitmap::createFromImageDraw(*nativeImage, DestinationColorSpace::SRGB(), size, ShareableBitmap::ShouldShrinkToFit::Yes)) {
```
Unless I'm missing something this refactor moves from vector scaling to bitmap scaling. We should probably continue to use `SVGImageForContainer`. Anyone ideas for how to add test coverage for that?

szewai replied:
I updated the patch to use `SVGImageForContainer` (in new function `nativeImageFromSVGImage`).
Also updated the test to ensure image data is expected.

shallawa commented on Source/WebCore/platform/graphics/ImageUtilities.h:
```diff
@@ -56,7 +56,8 @@ enum class ImageDecodingError : uint8_t {
 WEBCORE_EXPORT String descriptionString(ImageDecodingError);
 WEBCORE_EXPORT Expected<std::pair<String, Vector<IntSize>>, ImageDecodingError> utiAndAvailableSizesFromImageData(std::span<const uint8_t>);
-WEBCORE_EXPORT void createIconDataFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths, CompletionHandler<void(RefPtr<SharedBuffer>&&)>&&);
+WEBCORE_EXPORT void getBitmapsForIconFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths, CompletionHandler<void(Vector<RefPtr<ShareableBitmap>>&&)>&&);
```
How about `createBitmapsFromImageData()`?

1. We do not like to use the verb `get` in the function names. Besides `create` means you are returning something that does not exist.
2. The function is creating different `ShareableBitmaps` form different lengths. It should not be specific for`Icon`.

szewai replied:
> The function is creating different ShareableBitmaps form different lengths. It should not be specific forIcon.

This function sets `lengths` to supported ico image size if `lengths` is empty; I will move the setting to `WebPageProxy::createIconDataFromImageData` and make suggested changes for this function.

shallawa commented on Source/WebCore/platform/graphics/ShareableBitmap.cpp:
```diff
@@ -109,7 +109,7 @@ RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image,
 }
 
-RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image, const DestinationColorSpace& colorSpace, const IntSize& imageSize)
+RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image, const DestinationColorSpace& colorSpace, const IntSize& imageSize, ShouldShrinkToFit shouldShrinkToFit)
```
I do not think this `shouldShrinkToFit ` is needed. The caller can just call the above function:

```
RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image, const DestinationColorSpace& colorSpace)
{
    return createFromImageDraw(image, colorSpace, image.size());
}
```
It will do the same thing.

szewai replied:
hmm I cannot use that because I need to scale an image to a target size and draw.
To do that we need to specify the source size (the original image size) and target size to `drawNativeImage` (e.g. an image has size { 10, 10 } and target size is { 20, 20 }, then I need to pass both sizes to `drawNativeImage `).
The existing implementation of `createFromImageDraw` only passes one size to `drawNativeImage`.
Maybe I misunderstood your proposal.

shallawa replied:
Oh I see. In this case please use the same pattern which is used in `GraphicsContext::drawImage()`
```
RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image, const DestinationColorSpace& colorSpace)
{
    return createFromImageDraw(image, colorspace, image.size());
}

RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image, const DestinationColorSpace& colorSpace, const IntSize& destination)
{
    return createFromImageDraw(image, colorSpace, destination, destination);
}

RefPtr<ShareableBitmap> ShareableBitmap::createFromImageDraw(NativeImage& image, const DestinationColorSpace& colorSpace, const IntSize& destination,  const IntSize& source)
{
    auto bitmap = ShareableBitmap::create({ destination, colorSpace });
    ...
    context->drawNativeImage(image, FloatRect({ }, destination), FloatRect({ }, source), { CompositeOperator::Copy });
    return bitmap;
}
```

szewai replied:
Sure, updated; thanks for review!

annevk commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
 }
 
-static RefPtr<SharedBuffer> expandSVGImageToData(SVGImage& svgImage, ASCIILiteral uti, std::span<const unsigned> lengths)
+static RefPtr<NativeImage> nativeImageFromSVGImage(SVGImage& image, unsigned length)
```
I wonder if we want to be clearer here that this always returns a square/icon. But since it's a static and pretty obvious it doesn't matter too much I think.

shallawa replied:
`nativeImageFromSVGImage ` -> `createNativeImageFromSVGImage `?

Also can the caller pass (..., const FloatSize& size) instead of passing `(...,unsigned length)`?

annevk reviewed (approved):
This looks good to me. Ideally @shallawa takes a look as well though.

Unfortunately I'm not really sure how to make the new tests cross-platform, but I suppose worst case we could only run those asserts on macOS.

sheeparegreat reviewed (approved):
Looks good to me. Return flow type is `Vector<RefPtr<ShareableBitmap>>`

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-    RetainPtr destination = adoptCF(CGImageDestinationCreateWithData(destinationData.get(), cfUTI.get(), lengths.size(), nullptr));
+    Vector<RefPtr<ShareableBitmap>> bitmaps;
+    auto sourceColorSpace = image.colorSpace();
+    auto destinationColorSpace = sourceColorSpace.supportsOutput() ? sourceColorSpace : DestinationColorSpace::SRGB();
```
If the NativeImage is HDR and its colorspace does not support output, the HDR contents will be lost by choosing SRGB.

szewai replied:
I guess this is probably not so important for the specific case (icon); I will add a comment here.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-            return nullptr;
-
-        CGImageDestinationAddImage(destination.get(), newImage.get(), nullptr);
+        if (RefPtr bitmap = ShareableBitmap::createFromImageDraw(image, destinationColorSpace, { (int)length, (int)length }, ShareableBitmap::ShouldShrinkToFit::Yes)) {
```
Again this interface is a little bit weird. You are passing the target size of the new image but you are telling `ShareableBitmap::createFromImageDraw()` also to use this size not the image size.

szewai replied:
This is because we need scaling at the draw so we need both sizes (image size and target size); I tried passing one size and it didn't work (with the current implementation of `createFromImageDraw` that only passes one size to `drawNativeImage`).

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-        RefPtr buffer = ImageBuffer::create(size, RenderingMode::Unaccelerated, RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), ImageBufferPixelFormat::BGRA8);
-        if (!buffer)
-            return nullptr;
+    FloatSize size { float(length), float(length) };
```
I think there are several ways to avoid this casting:
```
FloatSize size(length, length);
IntSize size(length, length); //  ImageBuffer::create() will do the casting from IntSize to FloatSize
FloatSize size { IntSize { length, length } };
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-        if (!buffer)
-            return nullptr;
+    FloatSize size { float(length), float(length) };
+    RefPtr buffer = ImageBuffer::create(size, RenderingMode::Unaccelerated, RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), ImageBufferPixelFormat::BGRA8);
```
Using SRGB will remove all HDR contents if they exist. This is not specific to this patch. SRGB is used everywhere and we should revisit of all of them. So you not have to fix it in this PR.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
 }
 
-static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILiteral uti, std::span<const unsigned> lengths)
+static Vector<RefPtr<ShareableBitmap>> bitmapsFromNativeImage(NativeImage& image, std::span<const unsigned> lengths)
```
`bitmapsFromNativeImage` -> `createBitmapsFromNativeImage`?

shallawa replied:
Also I think this function should return `Vector<Ref<ShareableBitmap>>` since we will never add a nullptr in the returned `Vector`.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+}
 
-        CGImageDestinationAddImage(destination.get(), newImage->platformImage().get(), nullptr);
+static Vector<RefPtr<ShareableBitmap>> bitmapsFromSVGImage(SVGImage& image, std::span<const unsigned> lengths)
```
`bitmapsFromSVGImage` -> `createBitmapsFromSVGImage`?

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+            }
+        }
+        return { };
     }
```
Early returns will make this code more readable. It took sometimes to realize the `continue; ` is there just to avoid going through `return { };`

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
 }
 
-void createIconDataFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths, CompletionHandler<void(RefPtr<SharedBuffer>&&)>&& completionHandler)
+void getBitmapsForIconFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths, CompletionHandler<void(Vector<RefPtr<ShareableBitmap>>&&)>&& completionHandler)
```
`getBitmapsForIconFromImageData` -> `createBitmapsFromImageData`?

With this PR, nothing in this function is related to `Icon`.

shallawa commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1541,8 +1541,11 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+        if (bitmaps.isEmpty())
+            return completionHandler(nullptr);
+
+        completionHandler(createIconDataFromBitmaps(WTFMove(bitmaps)));
```
Let's stick with one verb `create` or `get` so the calling sequence is easy to follow:
```
createIconDataFromImageData()
    createBitmapsFromImageData()
        createBitmapsFromSVGImage()
            createNativeImageFromSVGImage()
    createIconDataFromBitmaps()
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+
+    for (auto bitmap : bitmaps) {
+        if (!bitmap)
+            return nullptr;
```
This will not be needed if this functions takes `Vector<Ref<ShareableBitmap>>`.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
+    RetainPtr destination = adoptCF(CGImageDestinationCreateWithData(destinationData.get(), cfUTI.get(), bitmaps.size(), nullptr));
+
+    for (auto bitmap : bitmaps) {
```
Should it `RefPtr` instead of `auto`?

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+
+        RetainPtr cgImage = bitmap->makeCGImageCopy();
+        if (!cgImage)
+            return nullptr;
```
Because makeCGImageCopy() can return nullptr, should we log some information about this image? At least we can log the size of the `ShareableBitmap`:
```
        LOG_WITH_STREAM(Images, stream
            << "createIconDataFromBitmaps() failed ShareableBitmap::makeCGImageCopy()"
            << " bitmap size: " << bitmap->size());
```

shallawa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -266,10 +268,30 @@
+        return nullptr;
+
+    WebCore::ImageAdapter adapter { *bitmapImage };
+    return bridge_cast(adapter.tiffRepresentation());
```
I don't think you need to create a `BitmapImage` or `ImageAdapter` here. You can use the static method `ImageAdapter::tiffRepresentation()`
```
WebCore::ImageAdapter::tiffRepresentation({ WebCore::NativeImage::create(WTFMove(cgImage)) });
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,97 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-        RetainPtr context = adoptCF(CGBitmapContextCreate(nullptr, length, length, 8, length * 4, colorSpace.get(), kCGImageAlphaPremultipliedLast));
-        if (!context)
-            return nullptr;
+        RefPtr bitmap = ShareableBitmap::createFromImageDraw(image, destinationColorSpace, { (int)length, (int)length }, ShareableBitmap::ShouldShrinkToFit::Yes);
```
This can be written like this (See above):
``` 
RefPtr bitmap = ShareableBitmap::createFromImageDraw(image, destinationColorSpace, { (int)length, (int)length }, image.size());
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -206,82 +206,97 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
-        RefPtr newImage = ImageBuffer::sinkIntoNativeImage(WTFMove(buffer));
-        if (!newImage)
-            return nullptr;
+        RefPtr bitmap = ShareableBitmap::createFromImageDraw(*nativeImage, DestinationColorSpace::SRGB(), size, ShareableBitmap::ShouldShrinkToFit::Yes);
```
This can be written like this (See above):
``` 
RefPtr bitmap = ShareableBitmap::createFromImageDraw(*nativeImage, DestinationColorSpace::SRGB(), size, nativeImage->size());
```

shallawa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
```diff
@@ -32,9 +32,11 @@
 #import "TestNavigationDelegate.h"
 #import "TestWKWebView.h"
 #import <UniformTypeIdentifiers/UniformTypeIdentifiers.h>
+#import <WebCore/BitmapImage.h>
```
I think this header file is not needed.

szewai committed:
Make web process send ShareableBitmap instead of SharedBuffer for icon data
https://bugs.webkit.org/show_bug.cgi?id=290873
rdar://143579665

Reviewed by Said Abou-Hallawa and Anne van Kesteren.

In existing implementation of icon data generation, UI process sends image data to web process, and web process decodes
image from data, generates images with different sizes, combines them into one ico image and sends back image data to UI
process. To make the process more safe, this patch makes web process send images (bitmaps) with different sizes to UI
process, and UI process is responsible for combining them into an ico image.

* Source/WebCore/platform/graphics/ImageAdapter.h:
(WebCore::ImageAdapter::image const):
* Source/WebCore/platform/graphics/ImageUtilities.h:
* Source/WebCore/platform/graphics/ShareableBitmap.cpp:
(WebCore::ShareableBitmap::createFromImageDraw):
* Source/WebCore/platform/graphics/ShareableBitmap.h:
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::createBitmapsFromNativeImage):
(WebCore::createNativeImageFromSVGImage):
(WebCore::createBitmapsFromSVGImage):
(WebCore::createBitmapsFromImageData):
(WebCore::createIconDataFromBitmaps):
(WebCore::expandNativeImageToData): Deleted.
(WebCore::expandSVGImageToData): Deleted.
(WebCore::createIconDataFromImageData): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::createIconDataFromImageData):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::createBitmapsFromImageData):
(WebKit::WebPage::createIconDataFromImageData): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/icon-svg-16.tiff: Added.
* Tools/TestWebKitAPI/Tests/WebKit/icon-svg-256.tiff: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/LoadAndDecodeImage.mm:
(TestWebKitAPI::tiffRepresentation):
(TestWebKitAPI::TEST(WebKit, CreateIconDataFromImageDataSVG)):

Canonical link: https://commits.webkit.org/293430@main
